File: src/qubitcoin/aether/coreference.py
"""
Coreference Resolution for Multi-Turn Chat (#54)

Resolve pronouns and references in multi-turn Aether Tree conversations:
- Pronoun resolution (it, this, that, they, them, etc.)
- Domain-specific references (the contract, the block, the transaction)
- Most-recent-entity-of-matching-type resolution strategy
- Mention tracking across turns
"""
import re
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

from ..utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class Mention:
    """A mention of an entity in text."""
    text: str          # The surface form (e.g., "it", "the block")
    entity_id: str     # Resolved entity ID or key
    position: int      # Character position in text
    type: str          # Entity type (block, transaction, address, contract, etc.)

    def to_dict(self) -> dict:
        return {
            'text': self.text,
            'entity_id': self.entity_id,
            'position': self.position,
            'type': self.type,
        }


# Pronoun → expected entity types mapping
_PRONOUN_TYPES: Dict[str, List[str]] = {
    'it': ['block', 'transaction', 'contract', 'address', 'metric', 'token', 'concept'],
    'its': ['block', 'transaction', 'contract', 'address', 'metric', 'token', 'concept'],
    'this': ['block', 'transaction', 'contract', 'address', 'metric', 'topic', 'concept'],
    'that': ['block', 'transaction', 'contract', 'address', 'metric', 'topic', 'concept'],
    'they': ['address', 'node', 'peer', 'validator', 'agent'],
    'them': ['address', 'node', 'peer', 'validator', 'agent'],
    'their': ['address', 'node', 'peer', 'validator', 'agent'],
    'these': ['block', 'transaction', 'contract', 'node'],
    'those': ['block', 'transaction', 'contract', 'node'],
}

# Domain-specific references → expected entity types
_REFERENCE_PATTERNS: List[Tuple[str, str]] = [
    (r'\bthe\s+block\b', 'block'),
    (r'\bthe\s+transaction\b', 'transaction'),
    (r'\bthe\s+tx\b', 'transaction'),
    (r'\bthe\s+contract\b', 'contract'),
    (r'\bthe\s+address\b', 'address'),
    (r'\bthe\s+wallet\b', 'address'),
    (r'\bthe\s+miner\b', 'address'),
    (r'\bthe\s+node\b', 'node'),
    (r'\bthe\s+peer\b', 'peer'),
    (r'\bthe\s+token\b', 'token'),
    (r'\bthe\s+bridge\b', 'bridge'),
    (r'\bthe\s+chain\b', 'chain'),
    (r'\bthe\s+network\b', 'network'),
    (r'\bthe\s+reward\b', 'metric'),
    (r'\bthe\s+difficulty\b', 'metric'),
    (r'\bthe\s+phi\b', 'metric'),
    (r'\bthe\s+same\s+(\w+)\b', 'same_ref'),  # "the same block"
]


class CoreferenceResolver:
    """Resolve pronouns and references in multi-turn chat."""

    def __init__(self) -> None:
        self._entity_register: List[Dict[str, Any]] = []
        self._max_register: int = 200

        # Stats
        self._resolves: int = 0
        self._mentions_tracked: int = 0
        self._successful_resolutions: int = 0
        self._last_resolve: float = 0.0

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def resolve(self, text: str, context: List[Dict[str, Any]]) -> str:
        """Resolve pronouns and references in text using context.

        Args:
            text: The user's message text.
            context: List of dicts with entity info from previous turns.
                     Each dict should have: 'id' (str), 'type' (str),
                     'text' (str, display name), 'turn' (int, optional).

        Returns:
            Text with pronouns replaced by their referents (in brackets).
        """
        self._resolves += 1
        self._last_resolve = time.time()

        # Update entity register with new context
        for entity in context:
            self._register_entity(entity)

        if not self._entity_register:
            return text

        resolved = text
        replacements: List[Tuple[int, int, str]] = []

        # Resolve pronouns
        for pronoun, expected_types in _PRONOUN_TYPES.items():
            pattern = re.compile(
                r'\b' + re.escape(pronoun) + r'\b',
                re.IGNORECASE,
            )
            for match in pattern.finditer(text):
                referent = self._find_referent(expected_types)
                if referent:
                    self._successful_resolutions += 1
                    replacement = f"[{referent['text']}]"
                    replacements.append((match.start(), match.end(), replacement))

        # Resolve domain-specific references ("the block", "the contract")
        for pattern_str, entity_type in _REFERENCE_PATTERNS:
            if entity_type == 'same_ref':
                continue  # Handle separately
            pattern = re.compile(pattern_str, re.IGNORECASE)
            for match in pattern.finditer(text):
                referent = self._find_referent([entity_type])
                if referent:
                    self._successful_resolutions += 1
                    replacement = f"[{referent['text']}]"
                    replacements.append((match.start(), match.end(), replacement))

        # Resolve "the same X" references
        same_pattern = re.compile(r'\bthe\s+same\s+(\w+)\b', re.IGNORECASE)
        for match in same_pattern.finditer(text):
            ref_type = self._type_from_word(match.group(1))
            if ref_type:
                referent = self._find_referent([ref_type])
                if referent:
                    self._successful_resolutions += 1
                    replacement = f"[{referent['text']}]"
                    replacements.append((match.start(), match.end(), replacement))

        # Apply replacements in reverse order to preserve positions
        if replacements:
            # Remove overlapping replacements (keep first match)
            replacements.sort(key=lambda r: r[0])
            non_overlapping: List[Tuple[int, int, str]] = []
            last_end = -1
            for start, end, repl in replacements:
                if start >= last_end:
                    non_overlapping.append((start, end, repl))
                    last_end = end

            # Apply in reverse
            for start, end, repl in reversed(non_overlapping):
                resolved = resolved[:start] + repl + resolved[end:]

        return resolved

    def track_mentions(self, text: str,
                       entities: List[Dict[str, Any]]) -> List[Mention]:
        """Track entity mentions in text.

        Args:
            text: Text to scan for mentions.
            entities: Known entities with 'id', 'type', 'text' fields.

        Returns:
            List of Mention objects found in text.
        """
        mentions: List[Mention] = []
        text_lower = text.lower()

        for entity in entities:
            entity_text = str(entity.get('text', '')).lower()
            entity_id = str(entity.get('id', ''))
            entity_type = str(entity.get('type', 'unknown'))

            if not entity_text or len(entity_text) < 2:
                continue

            # Find all occurrences
            start = 0
            while True:
                idx = text_lower.find(entity_text, start)
                if idx == -1:
                    break
                mentions.append(Mention(
                    text=text[idx:idx + len(entity_text)],
                    entity_id=entity_id,
                    position=idx,
                    type=entity_type,
                ))
                start = idx + 1

            # Register for future coreference resolution
            self._register_entity(entity)

        self._mentions_tracked += len(mentions)
        return mentions

    def register_entities_from_turn(self, entities: Dict[str, Any],
                                     turn: int = 0) -> None:
        """Register entities from a processed turn for future resolution.

        Args:
            entities: Dict of entity_type -> entity_value from extraction.
            turn: Turn number for recency tracking.
        """
        for etype, evalue in entities.items():
            if isinstance(evalue, list):
                for item in evalue:
                    self._register_entity({
                        'id': str(item),
                        'type': etype,
                        'text': str(item),
                        'turn': turn,
                    })
            elif isinstance(evalue, dict):
                for k, v in evalue.items():
                    self._register_entity({
                        'id': str(v),
                        'type': etype,
                        'text': str(v),
                        'turn': turn,
                    })
            else:
                self._register_entity({
                    'id': str(evalue),
                    'type': etype,
                    'text': str(evalue),
                    'turn': turn,
                })

    def reset(self) -> None:
        """Clear the entity register for a new conversation."""
        self._entity_register = []

    def get_stats(self) -> dict:
        """Return runtime statistics."""
        return {
            'resolves': self._resolves,
            'mentions_tracked': self._mentions_tracked,
            'successful_resolutions': self._successful_resolutions,
            'entities_registered': len(self._entity_register),
            'last_resolve': self._last_resolve,
        }

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _register_entity(self, entity: Dict[str, Any]) -> None:
        """Add an entity to the register (most recent first)."""
        if not entity.get('id') or not entity.get('type'):
            return

        # Check for duplicates — update turn if same entity
        for existing in self._entity_register:
            if existing['id'] == entity['id'] and existing['type'] == entity['type']:
                existing['turn'] = entity.get('turn', existing.get('turn', 0))
                return

        self._entity_register.append(entity)

        # Bound register size
        if len(self._entity_register) > self._max_register:
            self._entity_register = self._entity_register[-self._max_register:]

    def _find_referent(self, expected_types: List[str]) -> Optional[Dict[str, Any]]:
        """Find the most recent entity matching any of the expected types.

        Strategy: most recent entity of matching type (last in register).
        """
        for entity in reversed(self._entity_register):
            if entity.get('type') in expected_types:
                return entity
        # Fallback: return most recent entity regardless of type
        if self._entity_register:
            return self._entity_register[-1]
        return None

    @staticmethod
    def _type_from_word(word: str) -> Optional[str]:
        """Map a natural language word to an entity type."""
        word_lower = word.lower()
        word_to_type = {
            'block': 'block',
            'transaction': 'transaction',
            'tx': 'transaction',
            'contract': 'contract',
            'address': 'address',
            'wallet': 'address',
            'miner': 'address',
            'node': 'node',
            'peer': 'peer',
            'token': 'token',
            'bridge': 'bridge',
            'chain': 'chain',
            'network': 'network',
            'reward': 'metric',
            'difficulty': 'metric',
            'one': None,  # "the same one" — too ambiguous
        }
        return word_to_type.get(word_lower)

File: aether-core/crates/aether-nlp/src/coreference.rs
//! Coreference Resolution — Resolve pronouns/references in multi-turn chat.
//!
//! Rule-based: track noun phrases, replace pronouns with most recent matching antecedent.

use serde::{Deserialize, Serialize};
use std::collections::VecDeque;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// A mention of an entity in text.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Mention {
    pub text: String,
    pub entity_id: String,
    pub position: usize,
    pub entity_type: String,
}

/// An entity registered for coreference tracking.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisteredEntity {
    pub id: String,
    pub entity_type: String,
    pub text: String,
    pub turn: u32,
}

/// Pronoun types and the entity categories they typically refer to.
fn pronoun_types(pronoun: &str) -> Option<&'static [&'static str]> {
    match pronoun {
        "it" | "its" => Some(&["block", "transaction", "contract", "address", "metric", "token", "concept"]),
        "this" | "that" => Some(&["block", "transaction", "contract", "address", "metric", "topic", "concept"]),
        "they" | "them" | "their" => Some(&["address", "node", "peer", "validator", "agent"]),
        "these" | "those" => Some(&["block", "transaction", "contract", "node"]),
        _ => None,
    }
}

/// Map a word to an entity type.
fn type_from_word(word: &str) -> Option<&'static str> {
    match word {
        "block" => Some("block"),
        "transaction" | "tx" => Some("transaction"),
        "contract" => Some("contract"),
        "address" | "wallet" | "miner" => Some("address"),
        "node" => Some("node"),
        "peer" => Some("peer"),
        "token" => Some("token"),
        "bridge" => Some("bridge"),
        "chain" => Some("chain"),
        "network" => Some("network"),
        "reward" | "difficulty" | "phi" => Some("metric"),
        _ => None,
    }
}

// ---------------------------------------------------------------------------
// CoreferenceResolver
// ---------------------------------------------------------------------------

/// Resolve pronouns and references in multi-turn chat.
pub struct CoreferenceResolver {
    entity_register: VecDeque<RegisteredEntity>,
    max_register: usize,
    resolves: u64,
    successful_resolutions: u64,
}

impl CoreferenceResolver {
    pub fn new() -> Self {
        Self {
            entity_register: VecDeque::new(),
            max_register: 200,
            resolves: 0,
            successful_resolutions: 0,
        }
    }

    /// Resolve pronouns and references in text using registered entities.
    ///
    /// Returns text with pronouns replaced by `[referent]`.
    pub fn resolve(&mut self, text: &str, context: &[RegisteredEntity]) -> String {
        self.resolves += 1;

        // Update register with new context
        for entity in context {
            self.register_entity(entity.clone());
        }

        if self.entity_register.is_empty() {
            return text.to_string();
        }

        let mut replacements: Vec<(usize, usize, String)> = Vec::new();
        let text_lower = text.to_lowercase();

        // Resolve pronouns
        let pronouns = ["it", "its", "this", "that", "they", "them", "their", "these", "those"];
        for pronoun in &pronouns {
            if let Some(expected_types) = pronoun_types(pronoun) {
                let search = format!(" {} ", pronoun);
                let mut start = 0;
                while let Some(pos) = text_lower[start..].find(&search) {
                    let actual_pos = start + pos + 1; // skip leading space
                    let end_pos = actual_pos + pronoun.len();
                    if let Some(ref_text) = self.find_referent(expected_types).map(|r| r.text.clone()) {
                        self.successful_resolutions += 1;
                        replacements.push((actual_pos, end_pos, format!("[{}]", ref_text)));
                    }
                    start = start + pos + search.len();
                }
                // Check at start of text
                if text_lower.starts_with(*pronoun)
                    && text.get(pronoun.len()..pronoun.len() + 1).map_or(false, |c| !c.chars().next().unwrap_or('a').is_alphanumeric())
                {
                    if let Some(ref_text) = self.find_referent(expected_types).map(|r| r.text.clone()) {
                        self.successful_resolutions += 1;
                        replacements.push((0, pronoun.len(), format!("[{}]", ref_text)));
                    }
                }
            }
        }

        // Resolve "the <entity_type>" references
        let reference_patterns = [
            ("the block", "block"),
            ("the transaction", "transaction"),
            ("the contract", "contract"),
            ("the address", "address"),
            ("the wallet", "address"),
            ("the node", "node"),
            ("the token", "token"),
        ];

        for &(pattern, entity_type) in &reference_patterns {
            if let Some(pos) = text_lower.find(pattern) {
                if let Some(ref_text) = self.find_referent(&[entity_type]).map(|r| r.text.clone()) {
                    self.successful_resolutions += 1;
                    replacements.push((pos, pos + pattern.len(), format!("[{}]", ref_text)));
                }
            }
        }

        // Remove overlapping replacements (keep first)
        replacements.sort_by_key(|r| r.0);
        let mut non_overlapping: Vec<(usize, usize, String)> = Vec::new();
        let mut last_end = 0;
        for (start, end, repl) in replacements {
            if start >= last_end {
                non_overlapping.push((start, end, repl));
                last_end = end;
            }
        }

        // Apply in reverse order
        let mut result = text.to_string();
        for (start, end, repl) in non_overlapping.into_iter().rev() {
            if start <= result.len() && end <= result.len() {
                result.replace_range(start..end, &repl);
            }
        }

        result
    }

    /// Register an entity for future coreference resolution.
    pub fn register_entity(&mut self, entity: RegisteredEntity) {
        if entity.id.is_empty() || entity.entity_type.is_empty() {
            return;
        }

        // Update if already exists
        for existing in self.entity_register.iter_mut() {
            if existing.id == entity.id && existing.entity_type == entity.entity_type {
                existing.turn = entity.turn;
                return;
            }
        }

        self.entity_register.push_back(entity);

        // Bound register size
        while self.entity_register.len() > self.max_register {
            self.entity_register.pop_front();
        }
    }

    /// Track entity mentions in text.
    pub fn track_mentions(&self, text: &str, entities: &[RegisteredEntity]) -> Vec<Mention> {
        let text_lower = text.to_lowercase();
        let mut mentions = Vec::new();

        for entity in entities {
            let entity_text_lower = entity.text.to_lowercase();
            if entity_text_lower.len() < 2 {
                continue;
            }
            let mut start = 0;
            while let Some(idx) = text_lower[start..].find(&entity_text_lower) {
                let actual_idx = start + idx;
                mentions.push(Mention {
                    text: text[actual_idx..actual_idx + entity_text_lower.len()].to_string(),
                    entity_id: entity.id.clone(),
                    position: actual_idx,
                    entity_type: entity.entity_type.clone(),
                });
                start = actual_idx + 1;
            }
        }

        mentions
    }

    /// Clear the entity register.
    pub fn reset(&mut self) {
        self.entity_register.clear();
    }

    /// Get resolution statistics.
    pub fn get_stats(&self) -> (u64, u64, usize) {
        (self.resolves, self.successful_resolutions, self.entity_register.len())
    }

    // Internal: find most recent entity matching any expected type.
    fn find_referent(&self, expected_types: &[&str]) -> Option<&RegisteredEntity> {
        for entity in self.entity_register.iter().rev() {
            if expected_types.contains(&entity.entity_type.as_str()) {
                return Some(entity);
            }
        }
        // Fallback: most recent entity
        self.entity_register.back()
    }
}

impl Default for CoreferenceResolver {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_entity(id: &str, etype: &str, text: &str) -> RegisteredEntity {
        RegisteredEntity {
            id: id.to_string(),
            entity_type: etype.to_string(),
            text: text.to_string(),
            turn: 0,
        }
    }

    #[test]
    fn test_resolve_pronoun() {
        let mut resolver = CoreferenceResolver::new();
        let context = vec![make_entity("blk_100", "block", "block #100")];
        let result = resolver.resolve("What happened in it today?", &context);
        assert!(result.contains("[block #100]"));
    }

    #[test]
    fn test_resolve_the_block() {
        let mut resolver = CoreferenceResolver::new();
        let context = vec![make_entity("blk_200", "block", "block #200")];
        let result = resolver.resolve("Tell me about the block", &context);
        assert!(result.contains("[block #200]"));
    }

    #[test]
    fn test_no_resolution_without_context() {
        let mut resolver = CoreferenceResolver::new();
        let result = resolver.resolve("Tell me about it", &[]);
        assert_eq!(result, "Tell me about it");
    }

    #[test]
    fn test_track_mentions() {
        let resolver = CoreferenceResolver::new();
        let entities = vec![make_entity("addr_1", "address", "the wallet")];
        let mentions = resolver.track_mentions("Check the wallet balance", &entities);
        assert_eq!(mentions.len(), 1);
        assert_eq!(mentions[0].entity_type, "address");
    }

    #[test]
    fn test_register_dedup() {
        let mut resolver = CoreferenceResolver::new();
        resolver.register_entity(make_entity("a", "block", "blk1"));
        resolver.register_entity(make_entity("a", "block", "blk1"));
        assert_eq!(resolver.entity_register.len(), 1);
    }
}
